pkg/smclient/client.go

```diff
@@ -0,0 +1,234 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package smclient
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+
+	"github.com/Peripli/service-manager-cli/pkg/errors"
+	"github.com/Peripli/service-manager-cli/pkg/httputil"
+	"github.com/Peripli/service-manager-cli/pkg/types"
+)
+
+// Client should be implemented by SM clients
+//go:generate counterfeiter . Client
+type Client interface {
+	RegisterPlatform(*types.Platform) (*types.Platform, error)
+	RegisterBroker(*types.Broker) (*types.Broker, error)
+	ListBrokers() (*types.Brokers, error)
+	ListPlatforms() (*types.Platforms, error)
+	DeleteBroker(string) error
+	DeletePlatform(string) error
+	UpdateBroker(string, *types.Broker) (*types.Broker, error)
+	UpdatePlatform(string, *types.Platform) (*types.Platform, error)
+}
+
+type serviceManagerClient struct {
+	config     *ClientConfig
+	httpClient *http.Client
+	headers    *http.Header
+}
+
+// NewClient returns new SM client
+func NewClient(config *ClientConfig) Client {
+	client := &serviceManagerClient{config: config, httpClient: &http.Client{}}
+	client.headers = &http.Header{}
+	client.headers.Add("Content-Type", "application/json")
+	if len(client.config.Token) > 0 {
+		client.headers.Add("Authorization", client.config.Token)
+	}
+
+	return client
+}
+
+// RegisterPlatform registers a platform in the service manager
+func (client *serviceManagerClient) RegisterPlatform(platform *types.Platform) (*types.Platform, error) {
+	requestBody, err := json.Marshal(platform)
+	if err != nil {
+		return nil, err
+	}
+
+	buffer := bytes.NewBuffer(requestBody)
+	response, err := client.call(http.MethodPost, "/v1/platforms", buffer)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode != http.StatusCreated {
+		return nil, errors.ResponseError{StatusCode: response.StatusCode}
+	}
+
+	var newPlatform *types.Platform
+	err = httputil.UnmarshalResponse(response, &newPlatform)
+	if err != nil {
+		return nil, err
+	}
+
+	return newPlatform, nil
+}
+
+// RegisterBroker registers a broker in the service manager
+func (client *serviceManagerClient) RegisterBroker(broker *types.Broker) (*types.Broker, error) {
+	requestBody, err := json.Marshal(broker)
+	if err != nil {
+		return nil, err
+	}
+
+	buffer := bytes.NewBuffer(requestBody)
+	response, err := client.call(http.MethodPost, "/v1/service_brokers", buffer)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode != http.StatusCreated {
+		return nil, errors.ResponseError{StatusCode: response.StatusCode}
+	}
+
+	var newBroker *types.Broker
+	err = httputil.UnmarshalResponse(response, &newBroker)
+	if err != nil {
+		return nil, err
+	}
+
+	return newBroker, nil
+}
+
+func (client *serviceManagerClient) ListBrokers() (*types.Brokers, error) {
+	brokers := &types.Brokers{}
+	err := client.list(brokers, "/v1/service_brokers")
+
```

Comment:
Is there a particular reason as to why the error is not checked here and instead it's directly returned to the calling function? (same question goes for the ListPlatforms() function)

I was thinking of the usual check that goes along like this:
```go
if err != nil {
    return nil, err
}

return brokers, nil
```

Reply:
It's simpler. 

The only difference is that it will return nil for platforms/brokers, the error will be the same. I don't see any benefit of checking the error here, if we are not enchanting it (the error) or short circuiting the execution of the method.
It is anyway better to return empty collection, than nil. The error will show what went wrong.